test/extensions/filters/network/common/fuzz/uber_filter.cc

```diff
@@ -0,0 +1,195 @@
+                                             Protobuf::Message* config_message) {
+  // System calls such as reading files are prohibited in this fuzzer. Some input that crashes the
+  // mock/fake objects are also prohibited.
+  const std::string name = Extensions::NetworkFilters::Common::FilterNameUtil::canonicalFilterName(
+      std::string(filter_name));
+  if (filter_name == NetworkFilterNames::get().DirectResponse) {
+    envoy::extensions::filters::network::direct_response::v3::Config& config =
+        dynamic_cast<envoy::extensions::filters::network::direct_response::v3::Config&>(
+            *config_message);
+    if (config.response().specifier_case() ==
+        envoy::config::core::v3::DataSource::SpecifierCase::kFilename) {
+      return true;
+    }
+  } else if (filter_name == NetworkFilterNames::get().LocalRateLimit) {
+    envoy::extensions::filters::network::local_ratelimit::v3::LocalRateLimit& config =
+        dynamic_cast<envoy::extensions::filters::network::local_ratelimit::v3::LocalRateLimit&>(
+            *config_message);
+    if (config.token_bucket().fill_interval().seconds() > seconds_in_one_day_) {
+      // Too large fill_interval may cause "c++/v1/chrono" overflow when simulated_time_system_ is
+      // converting it to a smaller unit. Constraining fill_interval to no greater than one day is
+      // reasonable.
+      return true;
+    }
+  }
+  return false;
+}
+
+void UberFilterFuzzer::fuzz(
+    const envoy::config::listener::v3::Filter& proto_config,
+    const Protobuf::RepeatedPtrField<::test::extensions::filters::network::Action>& actions) {
+  try {
+    // Try to create the filter callback(cb_). Exit early if the config is invalid or violates PGV
+    // constraints.
+    const std::string& filter_name = proto_config.name();
+    ENVOY_LOG_MISC(info, "filter name {}", filter_name);
+    auto& factory = Config::Utility::getAndCheckFactoryByName<
+        Server::Configuration::NamedNetworkFilterConfigFactory>(filter_name);
+    ProtobufTypes::MessagePtr message = Config::Utility::translateToFactoryConfig(
+        proto_config, factory_context_.messageValidationVisitor(), factory);
+    if (invalidInputForFuzzer(filter_name, message.get())) {
+      // Make sure no invalid system calls are executed in fuzzer.
+      return;
+    }
+    ENVOY_LOG_MISC(info, "Config content after decoded: {}", message->DebugString());
+    cb_ = factory.createFilterFactoryFromProto(*message, factory_context_);
+    perFilterSetup(proto_config.name());
+    // Add filter to connection_
+    cb_(read_filter_callbacks_->connection_);
+  } catch (const EnvoyException& e) {
+    ENVOY_LOG_MISC(debug, "Controlled exception in filter setup{}", e.what());
+    return;
+  }
+
+  for (const auto& action : actions) {
+    ENVOY_LOG_MISC(trace, "action {}", action.DebugString());
+    switch (action.action_selector_case()) {
+    case test::extensions::filters::network::Action::kOnData: {
+      if (read_filter_ != nullptr) {
```

Comment:
Why would read_filter_ be NULL?

Reply:
> Why would read_filter_ be NULL?

Oh, I thought filters' actions could be divided into two types: ReadFilter and WriteFilter. WriteFilter shouldn't call the method of ReadFilter and ReadFilter shouldn't call WriteFilter's method. 
Since the network filters that are currently supported are all ReadFilters, it is useless now. All current supported filters will trigger `ON_CALL(read_filter_callbacks_->connection_, addReadFilter(_))` so it is not NULL. But in the future, I will add a `write_filter_ ` amd maybe one of `read_filter_` or `write_filter_` will be NULL for a specific filter.

Do you have any suggestion on how to deal with these two types of filters? I think it would be more efficient to have two separate fuzzers for ReadFilter and Filter(which extends both). But we can also deal with them in one fuzzer.

Reply:
>  I think it would be more efficient to have two separate fuzzers for ReadFilter and Filter(which extends both). 

This might be nice, especially if the interface to fuzz looks very similar and there isn't much to port over besides the step to add the filter and the methods to fuzz. Having two separate filters makes sense for ReadFilter/WriteFilter, but having a fuzzer for Filter (for both), seems like it wouldn't do anything different, right?

Reply:
I agree that a fuzzer for ReadFilter, and a separate one for WriteFilter, makes sense.